daemon/cmd/devices.go

```diff
@@ -0,0 +1,257 @@
+// routes in the system.
+func (dm *DeviceManager) Detect() error {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+
+	if err := expandDevices(); err != nil {
+		log.WithError(err).Warnf("Failed to expand device wildcards")
+	}
+
+	// Detect the devices from the system routing table by picking the devices
+	// which have global unicast routes.
+	if len(option.Config.Devices) == 0 {
+		routes, err := netlink.RouteList(nil, netlink.FAMILY_ALL)
+		if err != nil {
+			return fmt.Errorf("cannot retrieve routes for device detection: %w", err)
+		}
+		dm.updateDevicesFromRoutes(routes)
+	} else {
+		for _, dev := range option.Config.Devices {
+			dm.devices[dev] = true
+		}
+	}
+
+	if option.Config.DirectRoutingDevice != "" {
+		dm.devices[option.Config.DirectRoutingDevice] = true
+	}
+
+	if option.Config.IPv6MCastDevice != "" {
+		dm.devices[option.Config.IPv6MCastDevice] = true
+	}
+
+	detectDirectRoutingDev := option.Config.EnableNodePort && option.Config.DirectRoutingDevice == ""
+	detectIPv6MCastDev := option.Config.EnableIPv6NDP && len(option.Config.IPv6MCastDevice) == 0
+	if detectDirectRoutingDev || detectIPv6MCastDev {
+		nodeIP := node.GetK8sNodeIP()
+
+		if nodeIP == nil {
+			return fmt.Errorf("Cannot determine direct routing or IPv6 multicast device as K8s node IP is not known")
+		} else {
+			for dev := range dm.devices {
+				link, err := netlink.LinkByName(dev)
+				if err != nil {
+					return fmt.Errorf("Cannot find device '%s': %w", dev, err)
+				}
+
+				// Check if any of the addresses assigned to this device is the k8s node IP and if so
+				// use it for direct routing and IPv6 multicast.
+				if addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL); err == nil {
+					for _, a := range addrs {
+						if a.IP.Equal(nodeIP) {
+							if detectDirectRoutingDev {
+								option.Config.DirectRoutingDevice = dev
+								log.Infof("Detected %s=%s", option.DirectRoutingDevice, option.Config.DirectRoutingDevice)
+							}
+							if detectIPv6MCastDev && link.Attrs().Flags&net.FlagMulticast != 0 {
+								option.Config.IPv6MCastDevice = dev
+								log.Infof("Detected %s=%s", option.IPv6MCastDevice, option.Config.IPv6MCastDevice)
+							}
+							break
+						}
+					}
+				} else {
+					log.WithError(err).Warnf("Cannot retrieve device '%s' IP addresses, skipping it", dev)
+				}
+			}
+
+			if detectDirectRoutingDev && option.Config.DirectRoutingDevice == "" {
+				return fmt.Errorf("Unable to determine BPF NodePort direct routing device. "+
+					"Use --%s to specify it", option.DirectRoutingDevice)
+			}
+			if detectIPv6MCastDev && option.Config.IPv6MCastDevice == "" {
+				return fmt.Errorf("Unable to determine Multicast device. Use --%s to specify them",
+					option.IPv6MCastDevice)
+			}
+		}
+	}
+
+	option.Config.Devices = dm.getDevices()
+	log.WithField(logfields.Devices, option.Config.Devices).Info("Detected devices")
+
+	return nil
+}
+
+func (dm *DeviceManager) GetDevices() []string {
```

Comment:
In a follow-up with runtime detection of devices the plan is to stop using `option.Config.Devices` throughout and rather call into device manager to ask for the list of devices as it may mutate over time. Hence I already have the state for the devices and this function here. 

See https://github.com/cilium/cilium/pull/17187 for the work in progress for the runtime detection and reasoning behind the structure of this file.